### PR TITLE
[DF] Make Aliases specific to the computation graph branch

### LINKED ISSUE #7381: [DF] Let Aliases be defined per computation graph branch, not globally
### Is your feature request related to a problem? Please describe.

While users can `Define` a column with the same name in different branches of the computation graph, that's not true for `Alias`es: they are global to the computation graph. For example, this fails:

```cpp
#include <ROOT/RDataFrame.hxx>
#include <iostream>

int main() {
   auto df = ROOT::RDataFrame(1).Define("x", [] { return 42; }).Define("y", [] { return 0; });
   auto dfzx = df.Alias("z", "x");
   auto dfzy = df.Alias("z", "y");
   auto max42 = dfzx.Max<int>("z");
   auto max0 = dfzy.Max<int>("z");
   std::cout << "should be 42: " << *max42 << '\n';
   std::cout << "should be 0: " << *max0 << '\n';

   return 0;
}
```

### Describe the solution you'd like

Aliases should behave the same as Defines, we should track them per-computation-graph-branch rather than per-computation-graph.

### Additional context

That would make it possible to `Redefine` aliases, which is currently not a thing because we don't want a computation-graph-branch-local `Redefine` to undefine a global `Alias`.

### PR DESCRIPTION
**Pull request by eguiraud**
As there is one RLoopManager per computation graph, when aliases
were managed by RLoopManager they were computation-graph-wide.
It is desirable to make Alias definitions behave coherently with
Defines instead, i.e. have Aliases be only accessible in the branch
of the computation graph in which they were defined, and only in
nodes that are downstream of the one where the alias is added.

This resolves #7381, "[DF] Let Aliases be defined per computation graph
branch, not globally".

In particular, in this commit:
- move alias management from RLoopManager to RBookedColumns
- remove alias-managing logic from RLoopManager
- refactor several functions so they only take RBookedColumns as input
  rather than a list of defined names plus a map of aliases
  (CheckForDefinition, CheckForRedefinition, FindUnknownColumns,
  GetValidatedColumnNames, BookFilterJit, FindUsedColumns,
  ParseRDFExpression)
- use RBookedColumns::ResolveAlias instead of ResolveAlias helper function
- adapt dataframe_utils test to the new signature of FindUnknownColumns

Note that this is a backward-incompatible change for user code that
relied on this global definition of Aliases. We expect this to be an
exceedingly rare usecase, and if such code exists it will produce a
clear error message.

### CODE REVIEW

**Review by Axel-Naumann — APPROVED** [review]
LGTM, just minor comments. Very nice improvement: it's indeed a more consistent, simpler design now! I am a bit worried that the pair `HasAlias()` / `ResolveAlias()` has a perf impact because the lookup needs to happen twice - but so far I see that mostly for errors where it shouldn't matter. An alternative might be to have `ResolveAlias()` return an empty string, if that can never be a valid column name, or to have it return a pair of "it's a valid alias, this is the aliased name".

File: tree/dataframe/inc/ROOT/RDF/RColumnRegister.hxx [hunk 1]
```diff
@@ -45,20 +45,32 @@ class RColumnRegister {
...
+
 private:
    /// Immutable map of Defines, can be shared among several nodes.
    /// When a new define is added (through a call to RInterface::Define or similar) a new map with the extra element is
    /// created.
    RDefineBasePtrMapPtr_t fDefines;
-   ColumnNamesPtr_t fColumnNames; ///< Names of Defines and Aliases registered so far.
+   /// Immutable map of Aliases, can be shared among several nodes.
+   std::shared_ptr<const std::unordered_map<std::string, std::string>> fAliases;
+   ColumnNamesPtr_t fColumnNames;  ///< Names of Defines and Aliases registered so far.
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RColumnRegister.hxx:63]: remnant of a clang-format pass on an half-baked revision, will fix thanks

File: tree/dataframe/src/RDFInterfaceUtils.cxx [hunk 2]
```diff
@@ -575,15 +564,14 @@ SelectColumns(unsigned int nRequiredNames, const ColumnNames_t &names, const Col
...
+                                 const RColumnRegister &definedCols, const ColumnNames_t &dataSourceColumns)
 {
    ColumnNames_t unknownColumns;
    for (auto &column : requiredCols) {
       const auto isBranch = std::find(datasetColumns.begin(), datasetColumns.end(), column) != datasetColumns.end();
       if (isBranch)
          continue;
-      const auto isDefine = std::find(definedCols.begin(), definedCols.end(), column) != definedCols.end();
-      if (isDefine)
+      if(definedCols.HasName(column))
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:574]: will run clang-format and fix this

File: tree/dataframe/src/RDFColumnRegister.cxx [hunk 3]
```diff
@@ -39,6 +39,36 @@ void RColumnRegister::AddName(std::string_view name)
...
+   return fAliases->find(name) != fAliases->end();
+}
+
+std::string RColumnRegister::ResolveAlias(std::string_view alias) const
+{
+   std::string aliasStr{alias};
+
+   // #var is an alias for R_rdf_sizeof_var
+   if (aliasStr.size() > 1 && aliasStr[0] == '#')
+      return "R_rdf_sizeof_" + aliasStr.substr(1);
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/src/RDFColumnRegister.cxx:63]: true but then

```cpp
"R_rdf_sizeof_" + aliasStr.substr(1)
```

would have to be

```cpp
std::string("R_rdf_sizeof_") + alias.substr(1)
```

and I don't see much advantage?

File: tree/dataframe/test/dataframe_utils.cxx [hunk 4]
```diff
@@ -137,7 +137,10 @@ TEST(RDataFrameUtils, FindUnknownColumns)
    TTree t("t", "t");
    t.Branch("a", &i);
 
-   auto ncols = RDFInt::FindUnknownColumns({"a", "b", "c", "d"}, RDFInt::GetBranchNames(t), {"b"}, {});
+   RDFInt::RColumnRegister defs;
+   defs.AddAlias("b", "a");
+
+   auto ncols = RDFInt::FindUnknownColumns({"a", "b", "c", "d"}, RDFInt::GetBranchNames(t), defs, {});
    EXPECT_STREQ("c", ncols[0].c_str());
    EXPECT_STREQ("d", ncols[1].c_str());
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/test/dataframe_utils.cxx:145]: done

File: tree/dataframe/test/dataframe_interface.cxx [hunk 5]
```diff
@@ -388,9 +388,10 @@ TEST(RDataFrameInterface, GetNSlots)
 TEST(RDataFrameInterface, DefineAliasedColumn)
 {
    ROOT::RDataFrame rdf(1);
-   auto r0 = rdf.Define("myVar", [](){return 1;});
+   auto r0 = rdf.Define("myVar", [] { return 1; });
    auto r1 = r0.Alias("newVar", "myVar");
-   EXPECT_ANY_THROW(r0.Define("newVar", [](int i){return i;}, {"myVar"})) << "No exception thrown when defining a column with a name which is already an alias.";
+   auto m = r0.Define("newVar", [] { return 42; }).Max<int>("newVar");
+   EXPECT_EQ(*m, 42);
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/test/dataframe_interface.cxx:394]: Not sure I understand the comment, but we are checking that the Define is allowed even if there is an Alias with the same name _in another branch of the graph_.

I updated the test so we now check the value of `newVar` in both branches.

File: tree/dataframe/test/dataframe_interface.cxx [hunk 6]
```diff
@@ -371,9 +388,12 @@ TEST(RDataFrameInterface, GetNSlots)
 TEST(RDataFrameInterface, DefineAliasedColumn)
 {
    ROOT::RDataFrame rdf(1);
-   auto r0 = rdf.Define("myVar", [](){return 1;});
+   auto r0 = rdf.Define("myVar", [] { return 1; });
    auto r1 = r0.Alias("newVar", "myVar");
-   EXPECT_ANY_THROW(r0.Define("newVar", [](int i){return i;}, {"myVar"})) << "No exception thrown when defining a column with a name which is already an alias.";
+   auto mdefine = r0.Define("newVar", [] { return 42; }).Max<int>("newVar");
+   EXPECT_EQ(*mdefine, 42);
+   auto malias = r1.Max<int>("newVar");
```
Comment (Axel-Naumann) [inline tree/dataframe/test/dataframe_interface.cxx:395]:
```suggestion
   auto malias = r1.Max<int>("newVar");
   EXPECT_EQ(*mdefine, 42);
```
for good style (run the event loop only once)?

Comment (eguiraud) [reply tree/dataframe/test/dataframe_interface.cxx:395]: haha yes! :smile:

**Pull request merged**